### ISSUE TITLE
Replication connection issue changes replication to a bad document

### ISSUE THREAD

**Issue by krainboltgreene** [apache/couchdb#2080]
I created a replication to a bad host and after failing it changed the replication entry to something that is malformed.

## Description

I created a replication from local database (`test`) to remote new database (`http://docker.for.mac.local:9292/test`). I gave it a hostname that was bad. After erroring out, the replication record changed to `localhost:5864/test` and now the edit page things I'm replicating locally but wont let me save any changes because it thinks I'm replicating to itself (invalid).

## Steps to Reproduce

1. Create a database with one record
2. Replicate to a bad hostname
3. Refresh the replication page

## Expected Behaviour

Failed replication that retries

## Your Environment

[TIP]:  # ( Include as many relevant details about your environment as possible. )
[TIP]:  # ( You can paste the output of curl http://YOUR-COUCHDB:5984/ here. )

* CouchDB Version used: latest docker
* Browser name and version: chrome latest
* Operating System and version: macos latest, otherwise docker latest

## Additional context

```
pouchdb_1   | [notice] 2019-07-23T04:17:33.305945Z nonode@nohost <0.1071.0> 435d4a6f97 localhost:5984 172.18.0.1 admin GET /_replicator/cf0cd4e6dd92cb3646d642115a002182 200 ok 10
pouchdb_1   | [notice] 2019-07-23T04:17:40.934630Z nonode@nohost <0.567.0> e1f2c8f2b4 0.0.0.0:5984 172.18.0.1 undefined GET /_active_tasks 401 ok 1
pouchdb_1   | [notice] 2019-07-23T04:17:50.155310Z nonode@nohost <0.1071.0> 2d74918a43 localhost:5984 172.18.0.1 admin GET /_utils/ 304 ok 3
pouchdb_1   | [notice] 2019-07-23T04:17:50.176553Z nonode@nohost <0.1071.0> 7fb61f43a3 localhost:5984 172.18.0.1 admin GET /_utils/dashboard.assets/css/styles.89dbadd64e69e0df7653.css 304 ok 4
pouchdb_1   | [notice] 2019-07-23T04:17:50.409778Z nonode@nohost <0.1071.0> a2acb86a0f localhost:5984 172.18.0.1 admin GET / 200 ok 1
pouchdb_1   | [notice] 2019-07-23T04:17:50.419012Z nonode@nohost <0.1071.0> 54ae075633 localhost:5984 172.18.0.1 admin GET / 200 ok 1
pouchdb_1   | [notice] 2019-07-23T04:17:50.421124Z nonode@nohost <0.1089.0> dfdf4d3978 localhost:5984 172.18.0.1 admin GET /_session 200 ok 1
pouchdb_1   | [notice] 2019-07-23T04:17:50.452824Z nonode@nohost <0.1089.0> cd8f437b03 localhost:5984 172.18.0.1 admin GET /_utils/dashboard.assets/img/couchdb-logo.png 304 ok 3
pouchdb_1   | [notice] 2019-07-23T04:17:50.631995Z nonode@nohost <0.1089.0> 1746fb6c92 localhost:5984 172.18.0.1 admin GET /_scheduler/jobs 200 ok 2
pouchdb_1   | [notice] 2019-07-23T04:17:50.642843Z nonode@nohost <0.1089.0> 92b92f6930 localhost:5984 172.18.0.1 admin GET /_all_dbs 200 ok 3
pouchdb_1   | [notice] 2019-07-23T04:17:50.706904Z nonode@nohost <0.1089.0> b74213fa41 localhost:5984 172.18.0.1 admin GET /_replicator/_all_docs?include_docs=true&limit=100 200 ok 6
pouchdb_1   | [notice] 2019-07-23T04:17:50.710850Z nonode@nohost <0.1071.0> c9e84c3cd4 localhost:5984 172.18.0.1 admin GET /_scheduler/docs?include_docs=true 200 ok 4
pouchdb_1   | [notice] 2019-07-23T04:17:50.729521Z nonode@nohost <0.1071.0> 1614d2270b localhost:5984 172.18.0.1 admin GET /_scheduler/jobs 200 ok 1
pouchdb_1   | [notice] 2019-07-23T04:17:50.812717Z nonode@nohost <0.1071.0> 3a6ffc9639 localhost:5984 172.18.0.1 admin GET /_utils/dashboard.assets/img/couchdb-logo.png 200 ok 8
pouchdb_1   | [notice] 2019-07-23T04:17:56.668162Z nonode@nohost <0.567.0> 7f83ecd0e9 0.0.0.0:5984 172.18.0.1 undefined GET /_active_tasks 401 ok 1
pouchdb_1   | [notice] 2019-07-23T04:18:13.465605Z nonode@nohost <0.567.0> 6b0773696a 0.0.0.0:5984 172.18.0.1 undefined GET /_active_tasks 401 ok 1
pouchdb_1   | [notice] 2019-07-23T04:18:19.895356Z nonode@nohost <0.370.0> -------- couch_replicator_scheduler: Job {"2813286e386a8db17c3aa9c7ac7222ad","+continuous"} started as <0.14881.0>
pouchdb_1   | [notice] 2019-07-23T04:18:24.532628Z nonode@nohost <0.14944.0> b91c094563 localhost:5984 127.0.0.1 undefined POST /_session 200 ok 1
pouchdb_1   | [notice] 2019-07-23T04:18:24.553092Z nonode@nohost <0.14944.0> 51a8b8cba1 localhost:5984 127.0.0.1 admin GET /test/ 200 ok 19
pouchdb_1   | [error] 2019-07-23T04:18:24.564399Z nonode@nohost <0.14881.0> -------- throw:{db_not_found,<<"could not open http://docker.for.mac.localhost:9292/test/">>}: Replication 2813286e386a8db17c3aa9c7ac7222ad+continuous failed to start "http://localhost:5984/test/" -> "http://docker.for.mac.localhost:9292/test/" doc <<"shards/a0000000-bfffffff/_replicator.1563854786">>:<<"cf0cd4e6dd92cb3646d642115a002182">> stack:[{couch_replicator_api_wrap,db_open,4,[{file,"src/couch_replicator_api_wrap.erl"},{line,114}]},{couch_replicator_scheduler_job,init_state,1,[{file,"src/couch_replicator_scheduler_job.erl"},{line,584}]}]
pouchdb_1   | [notice] 2019-07-23T04:18:26.306657Z nonode@nohost <0.567.0> cd1605cdb4 0.0.0.0:5984 172.18.0.1 undefined GET /_active_tasks 401 ok 1
pouchdb_1   | [notice] 2019-07-23T04:18:41.721737Z nonode@nohost <0.567.0> 20a548f7c3 0.0.0.0:5984 172.18.0.1 undefined GET /_active_tasks 401 ok 1
pouchdb_1   | [notice] 2019-07-23T04:18:56.980503Z nonode@nohost <0.567.0> bfbffda5d9 0.0.0.0:5984 172.18.0.1 undefined GET /_active_tasks 401 ok 1
pouchdb_1   | [notice] 2019-07-23T04:19:11.987051Z nonode@nohost <0.567.0> 78d94ce5f7 0.0.0.0:5984 172.18.0.1 undefined GET /_active_tasks 401 ok 1

```

**Comment by krainboltgreene** [comment]
<img width="1379" alt="Screen Shot 2019-07-22 at 9 23 44 PM" src="https://user-images.githubusercontent.com/334809/61682591-13abb400-acc7-11e9-8ee3-9a45d958e191.png">
<img width="651" alt="Screen Shot 2019-07-22 at 9 23 49 PM" src="https://user-images.githubusercontent.com/334809/61682592-13abb400-acc7-11e9-8c93-3395bdf0b742.png">
<img width="1439" alt="Screen Shot 2019-07-22 at 9 24 04 PM" src="https://user-images.githubusercontent.com/334809/61682593-13abb400-acc7-11e9-85a6-464ca0c18432.png">

**Comment by krainboltgreene** [comment]
This is actually superbly strange.

So it looks like as soon as I save the replication it reverts in the UI, but keeps the database name:

<img width="1312" alt="Screen Shot 2019-07-22 at 9 34 23 PM" src="https://user-images.githubusercontent.com/334809/61682973-a9940e80-acc8-11e9-80ba-81b617ec339a.png">

But it tries in good faith to make the request:

```
pouchdb_1   | [error] 2019-07-23T04:34:22.330657Z nonode@nohost <0.31327.0> -------- throw:{db_not_found,<<"could not open http://docker.for.mac.localhost:9292/remotedocker2/">>}: Replication 78d567c6df8482337dc256ad49baf7c3+continuous+create_target failed to start "http://localhost:5984/test/" -> "http://docker.for.mac.localhost:9292/remotedocker2/" doc <<"shards/60000000-7fffffff/_replicator.1563854786">>:<<"cf0cd4e6dd92cb3646d642115a0045de">> stack:[{couch_replicator_api_wrap,db_open,4,[{file,"src/couch_replicator_api_wrap.erl"},{line,114}]},{couch_replicator_scheduler_job,init_state,1,[{file,"src/couch_replicator_scheduler_job.erl"},{line,584}]}]
```

BUT in my server logs for the target I actually get requests!

```
               | 127.0.0.1 - - [22/Jul/2019:21:36:32 -0700] "PUT /remotedocker2/ " 404 - 0.0000
               | 127.0.0.1 - - [22/Jul/2019:21:36:32 -0700] "GET /remotedocker2/ " 404 - 0.0000
```

**Comment by wohali** [comment]
In CouchDB 2.x you MUST use the full URL to any database, even local. This is a known issue.

If you must, specify the database with `http://localhost:5984/` if necessary.

Definitely delete your target database before starting this again.

**Comment by krainboltgreene** [comment]
Did you respond to the wrong ticket? I am extremely confused.

**Comment by nickva** [comment]
"Local" replications in CouchDB 2.x don't do what most users would expect them to do. 

It is mostly a legacy feature which is going to be removed in the next major release of CouchDB.

See this documentation PR for more details: https://github.com/apache/couchdb-documentation/pull/423

I'd follow @wohali's advice (who did reply to the right issue, btw) to use remote urls for source and target endpoints.

The reason you see "localhost:..." is that it was a patch / hack to make "local" replications work by transforming endpoints like "db" to a full url by using the ip and port the listening CouchDB server. That kind of / sort of works but not always and is somewhat surprising behavior.

**Comment by krainboltgreene** [comment]
It feels like the information I gave wasn't read?

> In CouchDB 2.x you MUST use the full URL to any database, even local. This is a known issue.

I am using a full URL, see the error message where couchdb fails to connect to `"http://localhost:5984/test/" -> "http://docker.for.mac.localhost:9292/remotedocker2/"`

> If you must, specify the database with http://localhost:5984/ if necessary.

I did. In the logs it specifically mentions `remotedocker2` (unless you're talking about the source, which also details that I'm talking about `test` database).

> "Local" replications in CouchDB 2.x don't do what most users would expect them to do.

It's not a local replication? Unless you mean "local" in the form of "same machine", but even then from couchdb's perspective different network.

> ransforming endpoints like "db" to a full url by using the ip and port the listening CouchDB server.

That's not happening here.

That's why I'm confused at the response/closed ticket.

**Comment by wohali** [comment]
Hi there,

Sorry for the confusion. Selecting "Local database" in the dropdown *is* the problem. Your first screenshot shows you doing this: https://user-images.githubusercontent.com/334809/61682592-13abb400-acc7-11e9-8c93-3395bdf0b742.png

Let's rule out this as the source of the problem in two ways:

1. Re-try your replication using Fauxton, and picking "Remote database" for the source.
2. Re-try your replication using `curl` instead and POSTing a document to the `_replicator` database.

Do either of these still cause the problem?

**Comment by krainboltgreene** [comment]
Ah I see the confusion. It's only local AFTER the bug (if it is a bug). I
set it as a remote target initially.

**Comment by krainboltgreene** [comment]
(I'm assuming we're talking about the target portion)

**Comment by wohali** [comment]
@krainboltgreene Both the source and the target matter. You must specify both as URLs. In that screenshot I linked it shows the source as a "local" database. That's the problem Nick and I are highlighting as well.

Did you get a chance to try this with `curl` instead of Fauxton? If so, can you paste your full command line so we can check it? (Don't forget to censor your admin password.)

**Comment by krainboltgreene** [comment]
As per the body of the issue I did enter two urls. The screenshots I'm showing are AFTER the bug has taken place.

**Comment by wohali** [comment]
Thanks @krainboltgreene . How about trying with `curl`?

**Comment by krainboltgreene** [comment]
Will do, also compiling a much more complete list of screenshots for clarity.

**Comment by krainboltgreene** [comment]
Entirely fresh couchdb instance inside docker.

Here's my database with a single record:
<img width="1552" alt="Screen Shot 2019-07-24 at 1 12 02 PM" src="https://user-images.githubusercontent.com/334809/61825363-11f00680-ae15-11e9-9a58-418893bd9b90.png">
Showing that I have no replications:
<img width="1552" alt="Screen Shot 2019-07-24 at 1 12 41 PM" src="https://user-images.githubusercontent.com/334809/61825364-11f00680-ae15-11e9-84a2-94d7078afee4.png">
Here is the setup of the replication PRIOR TO SAVING:
<img width="1552" alt="Screen Shot 2019-07-24 at 1 14 23 PM" src="https://user-images.githubusercontent.com/334809/61825366-11f00680-ae15-11e9-9ded-c1cee0bb15d2.png">
The logs of me saving the replication and it starting:
<img width="1552" alt="Screen Shot 2019-07-24 at 1 14 36 PM" src="https://user-images.githubusercontent.com/334809/61825367-11f00680-ae15-11e9-92ea-9ff1748a8ea5.png">
The replication list AFTER I have saved (notice the target uri):
<img width="1552" alt="Screen Shot 2019-07-24 at 1 14 41 PM" src="https://user-images.githubusercontent.com/334809/61825368-12889d00-ae15-11e9-92da-f6a88e2bed7a.png">
The replication configuration of that replication (notice the malformed form):
<img width="1552" alt="Screen Shot 2019-07-24 at 1 14 49 PM" src="https://user-images.githubusercontent.com/334809/61825369-12889d00-ae15-11e9-9856-c80592396733.png">

**Comment by wohali** [comment]
Thanks, really need to see this from `curl` now, as I suspect a Fauxton bug.

**Comment by wohali** [comment]
OK, I know the problem.

DNS resolves all hostnames that end in `.localhost` to `127.0.0.1.`

Try it:

```
$ host foobar.localhost
foobar.localhost has address 127.0.0.1
foobar.localhost has IPv6 address ::1
```

This is why it's rewriting it as `127.0.0.1`.

Use a different hostname.

**Comment by krainboltgreene** [comment]
Fascinating! So then couchdb says "Well, now I'm doing local to local" and breaks the form?

Alright, if the fix is "find a better way to do this for development" then I can do that.

**Comment by krainboltgreene** [comment]
Although now that I think about it, it's weird that the UI/Form is incorrect despite the job still trying to reach docker.for.mac.localhost.

**Comment by wohali** [comment]
What Fauxton does is write the document, as you specify it, through to CouchDB.

It then reads that back when displaying the result.

CouchDB is rewriting the document, which is why you see what you see. :) And it does so in compliance with [RFC 2606](https://tools.ietf.org/html/rfc2606) so I'm marking this as a wontfix.

**Comment by krainboltgreene** [comment]
I guess if you say so, Fauxton seems to be really confused.

<img width="1429" alt="Screen Shot 2019-07-24 at 4 07 48 PM" src="https://user-images.githubusercontent.com/334809/61834557-496aad00-ae2d-11e9-9139-286062235bbc.png">

I appreciate your time investigating this issue deeply.

**Comment by wohali** [comment]
I don't have a great answer as to what Fauxton is doing here - if you really want that investigated deeply, please open a new issue here:  https://github.com/apache/couchdb-fauxton/issues

**Comment by nickva** [comment]
@wohali Good find! I had no idea about this behavior

@krainboltgreene Sorry for the confusion. The behavior was unexpected for me as well.

I did try it curl just to confirm it is Fauxton and not the backend.

```
{
  "database": "_replicator",
  "doc_id": "66c040ce5dfb2774eadd906008000326",
  "id": "e8e043603474d4b76abf6970ed458e92+continuous",
  "node": "node1@127.0.0.1",
  "source": "http://adm:*****@localhost:15984/source/",
  "target": "http://adm:*****@somename.localhost:15984/target/",
  "state": "crashing",
  "info": "{replication_auth_error,\n    {session_request_failed,\"http://somename.localhost:15984/_session\",\"adm\",\n        {conn_failed,{error,nxdomain}}}}",
  "error_count": 2,
  "last_updated": "2019-07-25T15:08:16Z",
  "start_time": "2019-07-25T15:06:58Z",
  "proxy": null
}
```

Also interestingly for me name.localhost doesn't resolve to 127.0.0.1, neither on a mac or in Debian VM in Vagrant.

**Comment by mrsarm** [comment]
I don't know how Vagrant VM works, but must be similar to Docker creating its own network environment, in my cases I was getting this `...{conn_failed,{error,nxdomain}` error trying to sync 2 Couch DBs, one running locally and the other in a Docker container, and turns out it was because `somename.localhost` added to my `/etc/hosts` is not available in my Docker containers (because it has its own `/etc/hosts`), so clearly the message says that the domain name cannot be resolved and it was because of that. Solution: use the local IP of your machine instead of a domain name, e.g. 192.168.0.34.
